File: typescript/Y2022/Day3.ts
import { Solution } from '../run';

const priority = (c: string): number => {
  if (c.toUpperCase() == c) {
    return c.charCodeAt(0) - 38;
  } else {
    return c.charCodeAt(0) - 96;
  }
};

const part1 = (input: string) => {
  let sum = 0;

  for (const line of input.trim().split('\n')) {
    const chars = line.trim().split('');
    const right = chars.slice(chars.length / 2);
    for (const c of chars) {
      if (right.includes(c)) {
        sum += priority(c);
        break;
      }
    }
  }

  return sum;
};

const part2 = (input: string) => {
  let sum = 0;

  const sacks = input
    .trim()
    .split('\n')
    .map((s) => new Set(s.trim().split('')));

  for (let i = 0; i < sacks.length; i++) {
    if (i % 3 == 0) {
      for (const c of sacks[i].values()) {
        if (sacks[i + 1].has(c) && sacks[i + 2].has(c)) {
          sum += priority(c);
          break;
        }
      }
    }
  }

  return sum;
};

const solution: Solution<number, number> = {
  part1,
  part2,
  answers: [7878, 2760],
};

export default solution;

File: src/y2022/day_3.rs
use crate::Solution;

pub fn get_solution() -> Solution<usize, usize> {
    Solution {
        part_1,
        part_2,
        answer: (7878, 2760),
    }
}

fn priority(c: char) -> usize {
    if c.is_uppercase() {
        c as usize - 38
    } else {
        c as usize - 96
    }
}

fn part_1(input: &str) -> anyhow::Result<usize> {
    let mut sum = 0;

    for line in input.trim().lines() {
        let chars = line.trim().chars();
        let middle = line.trim().len() / 2;
        let right: Vec<char> = chars.clone().skip(middle).collect();
        for c in chars.take(middle) {
            if right.contains(&c) {
                sum += priority(c);
                break;
            }
        }
    }

    Ok(sum)
}

fn part_2(input: &str) -> anyhow::Result<usize> {
    let mut sum = 0;

    let sacks = input
        .trim()
        .lines()
        .map(|s| s.trim().chars().collect::<Vec<_>>())
        .collect::<Vec<_>>();

    for (i, sack) in sacks.iter().enumerate().step_by(3) {
        for c in sack {
            if sacks[i + 1].contains(c) && sacks[i + 2].contains(c) {
                sum += priority(*c);
                break;
            }
        }
    }

    Ok(sum)
}
